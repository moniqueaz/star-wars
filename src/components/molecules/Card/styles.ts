import styled, { css } from 'styled-components';
import Text from 'components/atoms/Text';

export const Wrapper = styled.div`
  ${( { theme } ) => css`
    width: 100%;
    background-color: ${theme.colors.white};
  `}
`;

export const Top = styled.div`
  position: relative;
  margin-bottom: 3rem;
`;

export const List = styled.ul`
  ${( { theme } ) => css`
    color: ${theme.colors.primary};
    padding: ${theme.spacings.xxsmall};
    &:empty {
      display: none;
    }
  `}
`;

export const Item = styled.li`
  ${( { theme } ) => css`
    display: flex ;
    justify-content: flex-start;
    align-items: flex-start;
    &:empty {
      display: none;
    }
    & + li {
      margin-top: ${theme.spacings.xxsmall};
    }
  `}
`;

export const Label = styled.label`
  ${( { theme } ) => css`
    text-transform: capitalize;
    font-weight: ${theme.font.bold};
    
    &:after{
      content: ':';
      margin-right: 0.5rem;
    }
  `}
`;

export const Value = styled.span``;

export const TitleUI = styled(Text)`
  ${( { theme } ) => css`
    position: absolute;
    left: 50%;
    bottom: 0;
    transform: translate(-50%, 30%);
    color: ${theme.colors.primary};
    background-color: ${theme.colors.white};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};
    width: 80%;
    text-align: center;
    border-right: 3px solid ${theme.colors.primary};
    border-bottom: 3px solid ${theme.colors.primary};
  `}
`;
