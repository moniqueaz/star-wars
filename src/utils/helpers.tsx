export const fetcher = (url: string) => fetch(url).then(res => res.json());

export const getRandomInteger = (min: number, max: number) => {
  return Math.floor(Math.random() * (max - min)) + min;
};
