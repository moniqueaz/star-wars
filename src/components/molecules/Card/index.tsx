import * as S from './styles';
import Image from 'components/atoms/Image';

type ListProps = {[key: string]: string | string[] | number | boolean }

type CardProps = {
  image: string,
  title: string,
  list?: ListProps
}
const Card = ( { image, title, list = {} }: CardProps) => (
  <S.Wrapper>
    <S.Top>
      <Image src={image} size="full"/>
      <S.TitleUI color="dark" size="small">{title}</S.TitleUI>
    </S.Top>
    <S.List>
      {
        Object.keys(list).map((item: string) => <S.Item key={item}>
          <S.Item>
            <S.Label>{item.replace(/_/ig, ' ')}</S.Label>
            <S.Value>{list[item]}</S.Value>
          </S.Item>
        </S.Item>)
      }
    </S.List>
  </S.Wrapper>
);

export default Card;
