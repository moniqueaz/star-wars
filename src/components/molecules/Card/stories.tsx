import { Story, Meta } from '@storybook/react/types-6-0';
import Card from '.';

export default {
  title: 'Molecules/Card',
  component: Card,
} as Meta;

export const Default: Story = args => {
  return (
    <div style={{ maxWidth: '150px' }}>
      <Card
        {...args}
        image="https://via.placeholder.com/150"
        title="Luke Skywalker"
        list={{
          height: '172',
          mass: '77',
          hair_color: 'blond',
          skin_color: 'fair',
          eye_color: 'blue',
          birth_year: '19BBY',
          gender: 'male',
        }}
      />
    </div>
  );
};

export const SimpleCard: Story = () => {
  return (
    <div style={{ maxWidth: '150px' }}>
      <Card
        image="https://via.placeholder.com/150"
        title="People"
      />
    </div>
  );
};
