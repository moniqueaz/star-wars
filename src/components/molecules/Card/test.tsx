import { renderWithTheme } from 'utils/tests/helpers';
import { screen } from '@testing-library/react';

import Card from '.';

describe('<Card />', () => {
  it('should render', () => {
    const { container } = renderWithTheme(<Card image="/" title="Luke Skywalker"/>);
    expect(container.querySelector('ul')).toBeInTheDocument();
    expect(container.querySelector('li')).not.toBeInTheDocument();
    expect(container.querySelector('img')).toBeInTheDocument();
    expect(container.querySelector('p')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  } );

  it('should render correct title', () => {
    renderWithTheme(<Card image="/" title="Luke Skywalker"/>);
    expect(screen.getByText(/Luke Skywalker/i)).toBeInTheDocument();
  } );

  it('should render with list', () => {
    renderWithTheme(<Card image="/" title="Luke Skywalker" list={{
      height: '172',
      mass: '77',
      hair_color: 'blond',
      skin_color: 'fair',
      eye_color: 'blue',
      birth_year: '19BBY',
      gender: 'male',
    }}/>);
    expect(screen.getByText(/blond/i)).toBeInTheDocument();
  } );
} );
