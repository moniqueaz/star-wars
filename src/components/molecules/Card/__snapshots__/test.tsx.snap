// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render 1`] = `
.c3 {
  width: 100%;
  color: #1A4173;
  font-size: 1.4rem;
}

.c0 {
  width: 100%;
  background-color: #f1f1f1;
}

.c1 {
  position: relative;
  margin-bottom: 3rem;
}

.c5 {
  color: #1A4173;
  padding: 0.8rem;
}

.c5:empty {
  display: none;
}

.c4 {
  position: absolute;
  left: 50%;
  bottom: 0;
  -webkit-transform: translate(-50%,30%);
  -ms-transform: translate(-50%,30%);
  transform: translate(-50%,30%);
  color: #1A4173;
  background-color: #f1f1f1;
  padding: 0.8rem 1.6rem;
  width: 80%;
  text-align: center;
  border-right: 3px solid #1A4173;
  border-bottom: 3px solid #1A4173;
}

.c2 {
  width: 100%;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      class="c2"
      src="/"
    />
    <p
      class="c3 c4"
      color="dark"
    >
      Luke Skywalker
    </p>
  </div>
  <ul
    class="c5"
  />
</div>
`;
